// @remotion/studio/dist/helpers/timeline-layout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTimelineLayerHeight = exports.TIMELINE_LAYER_HEIGHT_AUDIO = exports.TIMELINE_LAYER_HEIGHT_IMAGE = exports.TIMELINE_LAYER_HEIGHT_VIDEO = exports.getExpandedTrackHeight = exports.getTreeRowHeight = exports.flattenVisibleTreeNodes = exports.buildTimelineTree = exports.getEffectSchemaLabels = exports.getSchemaFields = exports.EXPANDED_SECTION_PADDING_RIGHT = exports.EXPANDED_SECTION_PADDING_LEFT = exports.TREE_INDENT_PER_LEVEL = exports.TREE_GROUP_ROW_HEIGHT = exports.UNSUPPORTED_FIELD_ROW_HEIGHT = exports.SCHEMA_FIELD_ROW_HEIGHT = exports.TIMELINE_TRACK_EXPANDED_HEIGHT = exports.TIMELINE_ITEM_BORDER_BOTTOM = exports.TIMELINE_BORDER = exports.TIMELINE_PADDING = void 0;
exports.TIMELINE_PADDING = 16;
exports.TIMELINE_BORDER = 1;
exports.TIMELINE_ITEM_BORDER_BOTTOM = 1;
exports.TIMELINE_TRACK_EXPANDED_HEIGHT = 100;
exports.SCHEMA_FIELD_ROW_HEIGHT = 22;
exports.UNSUPPORTED_FIELD_ROW_HEIGHT = 22;
exports.TREE_GROUP_ROW_HEIGHT = 22;
exports.TREE_INDENT_PER_LEVEL = 16;
exports.EXPANDED_SECTION_PADDING_LEFT = 28;
exports.EXPANDED_SECTION_PADDING_RIGHT = 10;
const SUPPORTED_SCHEMA_TYPES = new Set([
    'number',
    'boolean',
    'rotation',
    'translate',
]);
const getSchemaFields = (controls) => {
    if (!controls) {
        return null;
    }
    return Object.entries(controls.schema).map(([key, fieldSchema]) => {
        const typeName = fieldSchema.type;
        const supported = SUPPORTED_SCHEMA_TYPES.has(typeName);
        return {
            key,
            description: fieldSchema.description,
            typeName,
            supported,
            rowHeight: supported
                ? exports.SCHEMA_FIELD_ROW_HEIGHT
                : exports.UNSUPPORTED_FIELD_ROW_HEIGHT,
            currentValue: controls.currentValue[key],
            fieldSchema,
        };
    });
};
exports.getSchemaFields = getSchemaFields;
const getEffectSchemaLabels = (effect) => {
    if (!effect.definition.schema) {
        return [];
    }
    return Object.entries(effect.definition.schema).map(([key, fieldSchema]) => ({
        key,
        description: fieldSchema.description,
    }));
};
exports.getEffectSchemaLabels = getEffectSchemaLabels;
const buildTimelineTree = (sequence) => {
    const roots = [];
    if (sequence.effects.length > 0) {
        roots.push({
            kind: 'group',
            id: `${sequence.id}::effects`,
            label: 'Effects',
            children: sequence.effects.map((effect, i) => {
                const effectId = `${sequence.id}::effects::${i}`;
                return {
                    kind: 'group',
                    id: effectId,
                    label: effect.definition.label,
                    children: (0, exports.getEffectSchemaLabels)(effect).map((label) => {
                        var _a;
                        return ({
                            kind: 'field',
                            id: `${effectId}::${label.key}`,
                            label: (_a = label.description) !== null && _a !== void 0 ? _a : label.key,
                            field: null,
                        });
                    }),
                };
            }),
        });
    }
    const controlFields = (0, exports.getSchemaFields)(sequence.controls);
    if (controlFields && controlFields.length > 0) {
        roots.push({
            kind: 'group',
            id: `${sequence.id}::controls`,
            label: 'Controls',
            children: controlFields.map((f) => {
                var _a;
                return ({
                    kind: 'field',
                    id: `${sequence.id}::controls::${f.key}`,
                    label: (_a = f.description) !== null && _a !== void 0 ? _a : f.key,
                    field: f,
                });
            }),
        });
    }
    return roots;
};
exports.buildTimelineTree = buildTimelineTree;
const flattenVisibleTreeNodes = (nodes, expandedTracks, depth = 0) => {
    var _a;
    const out = [];
    for (const node of nodes) {
        out.push({ node, depth });
        if (node.kind === 'group' && ((_a = expandedTracks[node.id]) !== null && _a !== void 0 ? _a : false)) {
            out.push(...(0, exports.flattenVisibleTreeNodes)(node.children, expandedTracks, depth + 1));
        }
    }
    return out;
};
exports.flattenVisibleTreeNodes = flattenVisibleTreeNodes;
const getTreeRowHeight = (node) => {
    if (node.kind === 'field' && node.field) {
        return node.field.rowHeight;
    }
    return exports.TREE_GROUP_ROW_HEIGHT;
};
exports.getTreeRowHeight = getTreeRowHeight;
const getExpandedTrackHeight = (sequence, expandedTracks) => {
    const tree = (0, exports.buildTimelineTree)(sequence);
    const flat = (0, exports.flattenVisibleTreeNodes)(tree, expandedTracks);
    if (flat.length === 0) {
        return exports.TIMELINE_TRACK_EXPANDED_HEIGHT;
    }
    const totalRowsHeight = flat.reduce((sum, { node }) => sum + (0, exports.getTreeRowHeight)(node), 0);
    const separators = Math.max(0, flat.length - 1);
    return totalRowsHeight + separators;
};
exports.getExpandedTrackHeight = getExpandedTrackHeight;
exports.TIMELINE_LAYER_HEIGHT_VIDEO = 75;
exports.TIMELINE_LAYER_HEIGHT_IMAGE = 50;
exports.TIMELINE_LAYER_HEIGHT_AUDIO = 25;
const getTimelineLayerHeight = (type) => {
    if (type === 'video') {
        return exports.TIMELINE_LAYER_HEIGHT_VIDEO;
    }
    if (type === 'image') {
        return exports.TIMELINE_LAYER_HEIGHT_IMAGE;
    }
    return exports.TIMELINE_LAYER_HEIGHT_AUDIO;
};
exports.getTimelineLayerHeight = getTimelineLayerHeight;
