// @remotion/studio/dist/components/Timeline/TimelineExpandedSection.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineExpandedSection = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const colors_1 = require("../../helpers/colors");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const ExpandedTracksProvider_1 = require("../ExpandedTracksProvider");
const TimelineExpandArrowButton_1 = require("./TimelineExpandArrowButton");
const TimelineFieldRow_1 = require("./TimelineFieldRow");
const TimelineListItem_1 = require("./TimelineListItem");
const expandedSectionBase = {
    color: 'white',
    fontFamily: 'Arial, Helvetica, sans-serif',
    fontSize: 12,
    display: 'flex',
    flexDirection: 'column',
    borderBottom: `1px solid ${colors_1.TIMELINE_TRACK_SEPARATOR}`,
};
const separator = {
    height: 1,
    backgroundColor: colors_1.TIMELINE_TRACK_SEPARATOR,
};
const groupRowBase = {
    height: timeline_layout_1.TREE_GROUP_ROW_HEIGHT,
    display: 'flex',
    alignItems: 'center',
    paddingRight: timeline_layout_1.EXPANDED_SECTION_PADDING_RIGHT,
};
const rowLabel = {
    fontSize: 12,
    color: 'rgba(255, 255, 255, 0.8)',
    userSelect: 'none',
};
const labelOnlyRowBase = {
    display: 'flex',
    alignItems: 'center',
    paddingRight: timeline_layout_1.EXPANDED_SECTION_PADDING_RIGHT,
};
const TimelineExpandedSection = ({ sequence, originalLocation, nestedDepth, nodePath }) => {
    var _a;
    var _b;
    const { expandedTracks, toggleTrack } = (0, react_1.useContext)(ExpandedTracksProvider_1.ExpandedTracksContext);
    const overrideId = (_b = (_a = sequence.controls) === null || _a === void 0 ? void 0 : _a.overrideId) !== null && _b !== void 0 ? _b : sequence.id;
    const validatedLocation = (0, react_1.useMemo)(() => {
        var _a;
        if (!originalLocation ||
            !originalLocation.source ||
            !originalLocation.line) {
            return null;
        }
        return {
            source: originalLocation.source,
            line: originalLocation.line,
            column: (_a = originalLocation.column) !== null && _a !== void 0 ? _a : 0,
        };
    }, [originalLocation]);
    const tree = (0, react_1.useMemo)(() => (0, timeline_layout_1.buildTimelineTree)(sequence), [sequence]);
    const flat = (0, react_1.useMemo)(() => (0, timeline_layout_1.flattenVisibleTreeNodes)(tree, expandedTracks), [tree, expandedTracks]);
    const expandedHeight = (0, react_1.useMemo)(() => (0, timeline_layout_1.getExpandedTrackHeight)(sequence, expandedTracks), [sequence, expandedTracks]);
    const sequenceOffsetPx = TimelineListItem_1.SPACING * 3 * nestedDepth;
    const keysToObserve = (0, react_1.useMemo)(() => {
        const fields = (0, timeline_layout_1.getSchemaFields)(sequence.controls);
        if (!fields) {
            return [];
        }
        return fields.map((f) => f.key);
    }, [sequence.controls]);
    const style = (0, react_1.useMemo)(() => {
        return {
            ...expandedSectionBase,
            height: expandedHeight,
        };
    }, [expandedHeight]);
    const renderRow = (0, react_1.useCallback)((node, depth) => {
        var _a;
        const paddingLeft = timeline_layout_1.EXPANDED_SECTION_PADDING_LEFT +
            depth * timeline_layout_1.TREE_INDENT_PER_LEVEL +
            sequenceOffsetPx;
        if (node.kind === 'group') {
            const isExpanded = (_a = expandedTracks[node.id]) !== null && _a !== void 0 ? _a : false;
            return (jsx_runtime_1.jsxs("div", { style: { ...groupRowBase, paddingLeft }, children: [
                    jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowButton, { isExpanded: isExpanded, onClick: () => toggleTrack(node.id), label: `${node.label} section`, hasExpandableContent: true }), jsx_runtime_1.jsx("span", { style: rowLabel, children: node.label })
                ] }));
        }
        if (node.field) {
            return (jsx_runtime_1.jsx(TimelineFieldRow_1.TimelineFieldRow, { field: node.field, overrideId: overrideId, validatedLocation: validatedLocation, paddingLeft: paddingLeft, nodePath: nodePath, keysToObserve: keysToObserve }));
        }
        return (jsx_runtime_1.jsx("div", { style: {
                ...labelOnlyRowBase,
                height: (0, timeline_layout_1.getTreeRowHeight)(node),
                paddingLeft,
            }, children: jsx_runtime_1.jsx("span", { style: rowLabel, children: node.label }) }));
    }, [
        expandedTracks,
        keysToObserve,
        nodePath,
        overrideId,
        sequenceOffsetPx,
        toggleTrack,
        validatedLocation,
    ]);
    if (flat.length === 0) {
        return jsx_runtime_1.jsx("div", { style: style, children: "No schema" });
    }
    return (jsx_runtime_1.jsx("div", { style: style, children: flat.map(({ node, depth }, i) => {
            return (jsx_runtime_1.jsxs(react_1.default.Fragment, { children: [i > 0 ? jsx_runtime_1.jsx("div", { style: separator }) : null, renderRow(node, depth)] }, node.id));
        }) }));
};
exports.TimelineExpandedSection = TimelineExpandedSection;
