// @remotion/studio/dist/components/Timeline/TimelineExpandArrowButton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineExpandArrowSpacer = exports.TimelineExpandArrowButton = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const arrowButton = {
    background: 'none',
    border: 'none',
    color: 'white',
    cursor: 'pointer',
    padding: 0,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    width: 12,
    height: 12,
    flexShrink: 0,
    fontSize: 8,
    marginRight: 4,
    userSelect: 'none',
    outline: 'none',
    lineHeight: 1,
};
const svgStyle = { display: 'block' };
const TimelineExpandArrowButton = ({ isExpanded, onClick, label, hasExpandableContent }) => {
    const style = (0, react_1.useMemo)(() => {
        return {
            ...arrowButton,
            transform: isExpanded ? 'rotate(90deg)' : 'rotate(0deg)',
            opacity: hasExpandableContent ? 1 : 0.2,
        };
    }, [isExpanded, hasExpandableContent]);
    return (jsx_runtime_1.jsx("button", { type: "button", style: style, onClick: onClick, disabled: !hasExpandableContent, "aria-expanded": isExpanded, "aria-label": `${isExpanded ? 'Collapse' : 'Expand'} ${label}`, children: jsx_runtime_1.jsx("svg", { width: "12", height: "12", viewBox: "0 0 8 8", style: svgStyle, children: jsx_runtime_1.jsx("path", { d: "M2 1L6 4L2 7Z", fill: "white" }) }) }));
};
exports.TimelineExpandArrowButton = TimelineExpandArrowButton;
const TimelineExpandArrowSpacer = () => {
    return jsx_runtime_1.jsx("div", { style: arrowButton });
};
exports.TimelineExpandArrowSpacer = TimelineExpandArrowSpacer;
