// @remotion/studio/dist/components/AudioWaveform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AudioWaveform = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const colors_1 = require("../helpers/colors");
const timeline_layout_1 = require("../helpers/timeline-layout");
const make_audio_waveform_worker_1 = require("../make-audio-waveform-worker");
const draw_peaks_1 = require("./draw-peaks");
const load_waveform_peaks_1 = require("./load-waveform-peaks");
const looped_media_timeline_1 = require("./looped-media-timeline");
const slice_waveform_peaks_1 = require("./slice-waveform-peaks");
const EMPTY_PEAKS = new Float32Array(0);
// Recreate the canvas after Fast Refresh because an already transferred canvas
// cannot be handed to OffscreenCanvas again.
const canRetryCanvasTransfer = (err) => {
    return err instanceof DOMException && err.name === 'InvalidStateError';
};
const canUseAudioWaveformWorker = () => {
    if (typeof Worker === 'undefined' ||
        typeof OffscreenCanvas === 'undefined' ||
        typeof HTMLCanvasElement === 'undefined') {
        return false;
    }
    return 'transferControlToOffscreen' in HTMLCanvasElement.prototype;
};
const container = {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    position: 'absolute',
    inset: 0,
};
const errorMessage = {
    fontSize: 13,
    paddingTop: 6,
    paddingBottom: 6,
    paddingLeft: 12,
    paddingRight: 12,
    alignSelf: 'flex-start',
    maxWidth: 450,
    opacity: 0.75,
};
const waveformCanvasStyle = {
    pointerEvents: 'none',
    width: '100%',
    height: '100%',
};
const volumeCanvasStyle = {
    position: 'absolute',
};
const drawLoopedWaveform = ({ canvas, peaks, volume, visualizationWidth, loopWidth, }) => {
    const h = canvas.height;
    const w = Math.ceil(visualizationWidth);
    const targetCanvas = document.createElement('canvas');
    targetCanvas.width = Math.max(1, Math.ceil(loopWidth));
    targetCanvas.height = h;
    (0, draw_peaks_1.drawBars)(targetCanvas, peaks, 'rgba(255, 255, 255, 0.6)', volume, targetCanvas.width);
    canvas.width = w;
    canvas.height = h;
    const ctx = canvas.getContext('2d');
    if (!ctx) {
        throw new Error('Failed to get canvas context');
    }
    const pattern = ctx.createPattern(targetCanvas, 'repeat-x');
    if (!pattern) {
        return;
    }
    pattern.setTransform(new DOMMatrix().scaleSelf(loopWidth / targetCanvas.width, 1));
    ctx.clearRect(0, 0, w, h);
    ctx.fillStyle = pattern;
    ctx.fillRect(0, 0, w, h);
};
const AudioWaveform = ({ src, startFrom, durationInFrames, visualizationWidth, volume, doesVolumeChange, playbackRate, loopDisplay, }) => {
    const [peaks, setPeaks] = (0, react_1.useState)(null);
    const [error, setError] = (0, react_1.useState)(null);
    const [waveformCanvasKey, setWaveformCanvasKey] = (0, react_1.useState)(0);
    const canUseWorkerPath = (0, react_1.useMemo)(() => canUseAudioWaveformWorker(), []);
    const vidConf = remotion_1.Internals.useUnsafeVideoConfig();
    if (vidConf === null) {
        throw new Error('Expected video config');
    }
    const containerRef = (0, react_1.useRef)(null);
    const waveformCanvas = (0, react_1.useRef)(null);
    const volumeCanvas = (0, react_1.useRef)(null);
    const waveformWorker = (0, react_1.useRef)(null);
    const hasTransferredCanvas = (0, react_1.useRef)(false);
    const latestRequestId = (0, react_1.useRef)(0);
    (0, react_1.useEffect)(() => {
        if (canUseWorkerPath) {
            return;
        }
        const controller = new AbortController();
        setPeaks(null);
        setError(null);
        (0, load_waveform_peaks_1.loadWaveformPeaks)(src, controller.signal)
            .then((p) => {
            if (!controller.signal.aborted) {
                setPeaks(p);
            }
        })
            .catch((err) => {
            if (!controller.signal.aborted) {
                setError(err);
            }
        });
        return () => controller.abort();
    }, [canUseWorkerPath, src]);
    (0, react_1.useEffect)(() => {
        if (!canUseWorkerPath) {
            return;
        }
        const canvasElement = waveformCanvas.current;
        if (!canvasElement || hasTransferredCanvas.current) {
            return;
        }
        const worker = (0, make_audio_waveform_worker_1.makeAudioWaveformWorker)();
        waveformWorker.current = worker;
        worker.addEventListener('message', (event) => {
            if (event.data.type === 'error') {
                if (event.data.requestId !== latestRequestId.current) {
                    return;
                }
                setError(new Error(event.data.message));
            }
        });
        let offscreen;
        try {
            offscreen = canvasElement.transferControlToOffscreen();
        }
        catch (err) {
            worker.terminate();
            waveformWorker.current = null;
            if (canRetryCanvasTransfer(err)) {
                setWaveformCanvasKey((key) => key + 1);
                return;
            }
            throw err;
        }
        hasTransferredCanvas.current = true;
        worker.postMessage({ type: 'init', canvas: offscreen }, [offscreen]);
        return () => {
            worker.postMessage({ type: 'dispose' });
            worker.terminate();
            waveformWorker.current = null;
            hasTransferredCanvas.current = false;
        };
    }, [canUseWorkerPath, waveformCanvasKey]);
    const portionPeaks = (0, react_1.useMemo)(() => {
        if (canUseWorkerPath || !peaks) {
            return null;
        }
        return (0, slice_waveform_peaks_1.sliceWaveformPeaks)({
            durationInFrames: (0, looped_media_timeline_1.shouldTileLoopDisplay)(loopDisplay)
                ? loopDisplay.durationInFrames
                : durationInFrames,
            fps: vidConf.fps,
            peaks,
            playbackRate,
            startFrom,
        });
    }, [
        canUseWorkerPath,
        durationInFrames,
        loopDisplay,
        peaks,
        playbackRate,
        startFrom,
        vidConf.fps,
    ]);
    (0, react_1.useEffect)(() => {
        const { current: canvasElement } = waveformCanvas;
        const { current: containerElement } = containerRef;
        if (!canvasElement || !containerElement) {
            return;
        }
        const h = containerElement.clientHeight;
        const w = Math.ceil(visualizationWidth);
        const vol = typeof volume === 'number' ? volume : 1;
        if (canUseWorkerPath) {
            const worker = waveformWorker.current;
            if (!worker || !hasTransferredCanvas.current) {
                return;
            }
            latestRequestId.current += 1;
            setError(null);
            const message = {
                type: 'render',
                requestId: latestRequestId.current,
                src,
                width: w,
                height: h,
                volume: vol,
                startFrom,
                durationInFrames,
                fps: vidConf.fps,
                playbackRate,
                loopDisplay,
            };
            worker.postMessage(message);
            return;
        }
        canvasElement.width = w;
        canvasElement.height = h;
        if ((0, looped_media_timeline_1.shouldTileLoopDisplay)(loopDisplay)) {
            drawLoopedWaveform({
                canvas: canvasElement,
                peaks: portionPeaks !== null && portionPeaks !== void 0 ? portionPeaks : EMPTY_PEAKS,
                volume: vol,
                visualizationWidth,
                loopWidth: (0, looped_media_timeline_1.getLoopDisplayWidth)({
                    visualizationWidth,
                    loopDisplay,
                }),
            });
        }
        else {
            (0, draw_peaks_1.drawBars)(canvasElement, portionPeaks !== null && portionPeaks !== void 0 ? portionPeaks : EMPTY_PEAKS, 'rgba(255, 255, 255, 0.6)', vol, w);
        }
    }, [
        canUseWorkerPath,
        durationInFrames,
        loopDisplay,
        playbackRate,
        portionPeaks,
        src,
        startFrom,
        vidConf.fps,
        visualizationWidth,
        volume,
        waveformCanvasKey,
    ]);
    (0, react_1.useEffect)(() => {
        const { current: volumeCanvasElement } = volumeCanvas;
        const { current: containerElement } = containerRef;
        if (!volumeCanvasElement || !containerElement) {
            return;
        }
        const h = containerElement.clientHeight;
        const context = volumeCanvasElement.getContext('2d');
        if (!context) {
            return;
        }
        volumeCanvasElement.width = Math.ceil(visualizationWidth);
        volumeCanvasElement.height = h;
        context.clearRect(0, 0, visualizationWidth, h);
        if (!doesVolumeChange || typeof volume === 'number') {
            return;
        }
        const volumes = volume.split(',').map((v) => Number(v));
        context.beginPath();
        context.moveTo(0, h);
        volumes.forEach((v, index) => {
            const x = (index / (volumes.length - 1)) * visualizationWidth;
            const y = (1 - v) * (h - timeline_layout_1.TIMELINE_BORDER * 2) + 1;
            if (index === 0) {
                context.moveTo(x, y);
            }
            else {
                context.lineTo(x, y);
            }
        });
        context.strokeStyle = colors_1.LIGHT_TRANSPARENT;
        context.stroke();
    }, [visualizationWidth, volume, doesVolumeChange]);
    if (error) {
        // eslint-disable-next-line no-console
        console.error(error);
        return (jsx_runtime_1.jsx("div", { style: container, children: jsx_runtime_1.jsx("div", { style: errorMessage, children: "No waveform available. Audio might not support CORS." }) }));
    }
    if (!canUseWorkerPath && !peaks) {
        return null;
    }
    return (jsx_runtime_1.jsxs("div", { ref: containerRef, style: container, children: [
            jsx_runtime_1.jsx("canvas", { ref: waveformCanvas, style: waveformCanvasStyle }, waveformCanvasKey), jsx_runtime_1.jsx("canvas", { ref: volumeCanvas, style: volumeCanvasStyle })
        ] }));
};
exports.AudioWaveform = AudioWaveform;
